mmedit/core/scheduler/lr_updater.py

```diff
@@ -50,3 +50,246 @@ def get_lr(self, runner, base_lr):
+                    for _regular_lr in self.regular_lr[k]
+                ]
+                lr_groups.update({k: _lr_group})
+            return lr_groups
+        else:
+            return [
+                self.get_lr(_regular_lr, 'generator')
+                for _regular_lr in self.regular_lr
+            ]
+
+    def _init_is_better(self, mode):
+        if mode == 'min':
+            self.mode_worse = float('inf')
+        else:
+            self.mode_worse = float('-inf')
+
+    def _reset(self):
+        self.best = self.mode_worse
+        self.cooldown_counter = 0
+        self.num_bad_epochs = 0
+
+    def is_better(self, a, best):
+        if self.mode == 'min' and self.threshold_mode == 'rel':
+            rel_epsilon = 1. - self.threshold
+            return a < best * rel_epsilon
+        elif self.mode == 'min' and self.threshold_mode == 'abs':
+            return a < best - self.threshold
+        elif self.mode == 'max' and self.threshold_mode == 'rel':
+            rel_epsilon = 1. + self.threshold
+            return a > best * rel_epsilon
+        else:
+            return a > best + self.threshold
+
+    @property
+    def in_cooldown(self):
+        return self.cooldown_counter > 0
+
+    def after_train_epoch(self, runner):
+        if not self.by_epoch:
+            return
+        cur_epoch = runner.epoch
+        if self.warmup is not None and self.warmup_by_epoch:
+            if cur_epoch <= self.warmup_epochs:
+                return
+        if self.val_metric is None:
+            current = runner.outputs['log_vars']['loss']
+            if self.is_better(current, self.best):
+                self.best = current
+                self.num_bad_epochs = 0
+            else:
+                self.num_bad_epochs += 1
+
+            if self.in_cooldown:
+                self.cooldown_counter -= 1
+                self.num_bad_epochs = 0
+            print(f'train_epoch--current {current:.6f} best {self.best:.6f}, '
+                  f'num_bad_epochs {self.num_bad_epochs}, '
+                  f'cooldown {self.in_cooldown} {self.cooldown_counter}')
+
+    def after_train_iter(self, runner):
+        if self.by_epoch:
+            return
+        cur_iter = runner.iter
+        if self.warmup_epochs is not None and cur_iter <= self.warmup_iters:
+            return
+        if self.val_metric is None:
+            current = runner.outputs['log_vars']['loss']
+            if self.is_better(current, self.best):
+                self.best = current
+                self.num_bad_epochs = 0
+            else:
+                self.num_bad_epochs += 1
+
+            if self.in_cooldown:
+                self.cooldown_counter -= 1
+                self.num_bad_epochs = 0
+            print(f'train_iter--current {current:.6f} best {self.best:.6f}, '
+                  f'num_bad_epochs {self.num_bad_epochs}, '
+                  f'cooldown {self.in_cooldown} {self.cooldown_counter}')
```

Comment:
If `val_metric` is specified, we check its value after training iteration based on **training loss**, else, we check it after validation iteration based on some **validation metrics**. Is it right?

We should make this clear in the docstring, otherwise, it's confusing, because it may engage in different hooks.